src/sys/petsc_interface.cxx

```diff
@@ -0,0 +1,248 @@
+    for (int z = 0; z < localmesh->LocalNz; z++) {
+      if (it.ind == localmesh->xstart) indices3D(it.ind - 1, localmesh->yend + 1, z) = counter++;
+      if (it.ind == localmesh->xend) indices3D(it.ind + 1, localmesh->yend + 1, z) = counter++;
+      indices3D(it.ind, localmesh->yend + 1, z) = counter++;
+    }
+  }
+  BOUT_FOR(i, localmesh->getRegion3D("RGN_NOY")) {
+    if ((i.x() >= localmesh->xstart && i.x() <= localmesh->xend) ||
+	(i.x() < localmesh->xstart && localmesh->firstX()) ||
+	(i.x() > localmesh->xend && localmesh->lastX())) {
+      indices3D[i] = counter++;
+    }
+  }
+
+  // Set up the 2D indices
+  counter = localmesh->globalStartIndex2D();
+  for (RangeIterator it=localmesh->iterateBndryLowerY(); !it.isDone(); it++) {
+      if (it.ind == localmesh->xstart) indices2D(it.ind - 1, localmesh->ystart - 1) = counter++;
+      if (it.ind == localmesh->xend) indices2D(it.ind + 1, localmesh->ystart - 1) = counter++;
+    indices2D(it.ind, localmesh->ystart - 1) = counter++;
+  }
+  for (RangeIterator it=localmesh->iterateBndryUpperY(); !it.isDone(); it++) {
+      if (it.ind == localmesh->xstart) indices2D(it.ind - 1, localmesh->yend + 1) = counter++;
+      if (it.ind == localmesh->xend) indices2D(it.ind + 1, localmesh->ystart + 1) = counter++;
+    indices2D(it.ind, localmesh->yend + 1) = counter++;
+  }
+  BOUT_FOR(i, localmesh->getRegion2D("RGN_NOY")) {
+    if ((i.x() >= localmesh->xstart && i.x() <= localmesh->xend) ||
+	(i.x() < localmesh->xstart && localmesh->firstX()) ||
+	(i.x() > localmesh->xend && localmesh->lastX())) {
+      indices2D[i] = counter++;
+    }
+  }
+
+  // Set up the Perp indices; will these work in general or will
+  // different ones be needed for each value of y?
+  counter = localmesh->globalStartIndexPerp();
+  BOUT_FOR(i, localmesh->getRegionPerp("RGN_NOY")) {
+    if ((i.x() >= localmesh->xstart && i.x() <= localmesh->xend) ||
+	(i.x() < localmesh->xstart && localmesh->firstX()) ||
+	(i.x() > localmesh->xend && localmesh->lastX())) {
+      indicesPerp[i] = counter++;
+    }
+  }
+}
+
+// PetscVectorElement implementation
+
+PetscVectorElement::PetscVectorElement(Vec* vector, int index) :
+  petscVector(vector), petscIndex(index) {
+}
+
+BoutReal PetscVectorElement::operator=(BoutReal val) {
+  auto status = VecSetValues(*petscVector, 1, &petscIndex, &val, INSERT_VALUES);
+  delete this;
+  if (status != 0) {
+    throw BoutException("Error when setting elements of a PETSc vector.");
+  }
+  return val;
+}
+
+BoutReal PetscVectorElement::operator+=(BoutReal val) {
+  auto status = VecSetValues(*petscVector, 1, &petscIndex, &val, ADD_VALUES);
+  delete this;
+  if (status != 0) {
+    throw BoutException("Error when setting elements of a PETSc vector.");
+  }
+  return val;
+}
+
+PetscVectorElement& PetscVectorElement::newElement(Vec* vector, PetscInt index) {
+  PetscVectorElement* ve = new PetscVectorElement(vector, index);
+  return *ve;
```

Comment:
I'm afraid I don't understand the need for this. Also looks like this leaks memory?
Why not just use the normal constructor?

Reply:
I want these objects to be destroyed immediately after they have been used for assignment. For this purpose I include `delete *this` towards the end of the assignment operators. In order to use `delete` I must be certain that the objects were created using the `new` keyword. Unfortunately, I can't just use a non-reference object because I can't assign to these when they are a function result on the LHS of an equals sign. In order to guarantee that the PetscVectorElement can be destroyed using `delete` I decided to restrict their creation to a method where I could be certain the `new` keyword was used. I concede it isn't a very nice approach so would be happy to hear alternatives.

Reply:
Take a look at [`std::bitset::operator[]`](https://en.cppreference.com/w/cpp/utility/bitset/operator_at) which does something similar for `std::bitset`.

You should be able to return a plain `PetscVectorElement` from `PetscVector::operator()`, although I suggest also making a `const` version that just returns a `BoutReal`.

Here's a possible implementation:

```cpp
template <class F>
class PetscVector {
public:
  using ind_type = typename F::ind_type;

  class Element {
  public:
     Element& operator=(BoutReal val) {
      VecSetValues(*petscVector, 1, &petscIndex, &val, INSERT_VALUES);
      return *this;
    }

    Element& operator+=(BoutReal val) {
      VecSetValues(*petscVector, 1, &petscIndex, &val, ADD_VALUES);
      return *this;
    }

    operator BoutReal() const {
      BoutReal value;
      VecGetValues(*petscVector, 1, &petscIndex, &value);
      return value;
    }

    Element() = delete;

    Element(const Element& other) = default;
    Element(Element&& other) = default;
    Element& operator=(const Element& rhs) = default;
    Element& operator=(Element&& rhs) = default;
    ~Element() = default;

    Element(Vec* vector, int index) : petscVector(vector), petscIndex(index) {}
  private:
    Vec* petscVector;
    PetscInt petscIndex;
  };
...
  BoutReal operator()(const ind_type& index) const {
    const int global = indexConverter->getGlobal(index);
    return Element(&vector, global);
  }

  Element operator()(ind_type& index) {
    const int global = indexConverter->getGlobal(index);
    return Element(&vector, global);
  }
  
...
};
```

(checking elided for clarity)

Here I've moved the class definition into `PetscVector`, which is not essential, but does make it a little cleaner. I've also `default`ed the copy/move constructors and assignment operators. They could be `delete`d, but I'm not sure I see the harm in allowing them.

We may need to provide the other arithmetic operators as well, if it makes sense.

Reply:
I've implemented this for the vector elements and confirmed that all tests compile and pass. I decided just not to put anything down for the copy/move constructors and assignments which (if I understand correctly) means the default implementations will be used. However, I'm wondering whether perhaps the assignments should be deleted or manually specified. We would want the code 
```C++
vectorObj(i) = vectorObj(j) = 5.;
```
to set both element i and j to 5, but if the default assignment is used then it would just set the ith Element to point at the jth Element without changing its value.

I'd like to similarly refactor the MatrixElement class, but things become a little unclear for the `operator BoutReal()` method. Because these "elements" may in fact correspond to multiple elements weighted for interpolation, there isn't really a single value which can be returned. The options I can see are:

1. Return the single i, j value anyway
2. Throw an error if the object is a setter for more than one element
3. Instead of a single value, return a vector of BoutReals

Reply:
As for the suggestion that we provide other arithmetic operators, I personally don't see a reason to, except possibly for `=-`, `=*`, and `=/`. Anything else would be handled by conversion to BoutReal, which would give the behaviour we want anyway. As for those other assignment operators, we could implement them but I'm not certain it's really necessary. I only provided `=` and `+=` because those correspond to the two types of setting which PETSc supports. I think the other setter operators are beyond the scope of my current project so don't plan to implement them myself.

Reply:
See discussion in PR #1805 .